notebook2/notebook/notebook.py

```diff
@@ -170,61 +198,201 @@ def external_url_for(path):
     return url + path
 
 
+# FIXME: use hash
+def UNSAFE_user_id_transform(user_id): return user_id.replace('|', '--_--')
+
+
+def read_svc_status(svc_name: str):
+    try:
+        # TODO: inspect exception for non-404
+        _ = k8s.read_namespaced_service(svc_name, 'default')
+        return 'Running'
+    except Exception:
+        return 'Deleted'
+
+
+def read_containers_status(container_statuses):
+    """
+        Summarize the container status based on its most recent state
+
+        Parameters
+        ----------
+        container_statuses : list[V1ContainerStatus]
+            https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1ContainerStatus.md
+    """
+    if container_statuses is None:
+        return None
+
+    state = container_statuses[0].state
+    rn = None
+    wt = None
+    tm = None
+    if state.running:
+        rn = {"started_at": state.running.started_at}
+
+    if state.waiting:
+        wt = {"reason": state.waiting.reason}
+
+    if state.terminated:
+        tm = {"exit_code": state.terminated.exit_code, "finished_at": state.terminated.finished_at,
+              "started_at": state.terminated.started_at, "reason": state.terminated.reason}
+
+    if rn is None and wt is None and tm is None:
+        return None
+
+    return {"running": rn, "terminated": tm, "waiting": wt}
+
+
+def read_conditions(conds):
+    """
+        Return the most recent status=="True" V1PodCondition or None
+
+        Parameters
+        ----------
+        conds : list[V1PodCondition]
+            https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1PodCondition.md
+
+    """
+    if conds is None:
+        return None
+
+    maxDate = None
+    maxCond = None
+    for condition in conds:
+        if maxDate is None:
+            maxCond = condition
+            maxDate = condition.last_transition_time
+            continue
+
+        if condition.last_transition_time > maxDate and condition.status == "True":
+            maxCond = condition
+            maxDate = condition.last_transition_time
+
+    return {
+        "message": maxCond.message,
+        "reason": maxCond.reason,
+        "status": maxCond.status,
+        "type": maxCond.type}
+
+
+def marshall_notebooks(pods = [], svc_status = None):
```

Comment:
Returns array because 1) in the future we may want to allow this 2) a deleted pod will not be cleared immediately, and we may either want the user to see this, somewhere (maybe behind a log dialog/slide-out element), or ignore such pods.